akka-stream-tests/src/test/scala/akka/stream/scaladsl/ActorRefSourceSpec.scala

```diff
@@ -129,6 +130,22 @@ class ActorRefSourceSpec extends StreamSpec {
       s.expectComplete()
     }
 
+    "complete and materialize the stream after receiving Status.Success" in assertAllStagesStopped {
+      val (ref, done) = {
+        Source.actorRef(3, OverflowStrategy.dropBuffer).toMat(Sink.ignore)(Keep.both).run()
+      }
+      ref ! Status.Success("ok")
```

Comment:
This can be solved by changing this line: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala#L49

to
```
if (bufferSize == 0 || buffer.isEmpty) onCompleteThenStop() // will complete the stream successfully
```

There is another context.stop in there also that I guess should also be changed.

Reply:
Good call; that fixed it indeed. I've swapped both of the out of place `context.stop`s with `onCompleteThenStop` and marked the poison pill test as ignore for the moment.

Should the documentation for `Source.actorRef` be updated here: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala#L473 to acknowledge the bug?

Reply:
Perhaps as an alternative to `PoisonPill` some other message (or message-matching function, like the `completionMatcher` function) should trigger the described behaviour (completing without first clearing the buffer). Then a caveat could be added explaining that simply `PoisonPill`ing the actor will not propagate a completion to the stream and should be avoided.

I suspect this probably worked at some point until some changes to the subscriber logic, though (which explains the left over `context.stop`s and the idea that killing the actor would inform the subscriber), so that's potentially an API change.

Reply:
It was probably a bad idea to try to support PoisonPill in the first place. We could add a warning that it doesn't work reliably and should be avoided. Instead use Success.

I doubt completing without draining is a feature that has been asked for, it's probably just a consequence of how PoisonPill works, so we don't have to replace that feature.